### DISCUSSION TITLE
Is it possible to change the type of the `message` field?

### DISCUSSION THREAD

**Discussion by winston0410** [colinhacks/zod#877] [Q&A] [answered]
Instead of a simple string, I would like to have a component returned with the error message when the error occurs. Is it possible to change the type of message? Maybe through a generic?

Some imaginary code right now:

```javascript
	const schema = z.object({
		'convert-characters': z
			.string()
			.nonempty({
                message: {
                    text: "The input field is empty right now.",
                    component: PlainText
                }
            })
			.refine((val) => {
				return false;
			}, {
            message: {
                text: "Containing invalid characters",
                component: WithLogic
            }
        })
	});
```

**Comment by JacobWeisenburger** [comment]
I would advise against this. From what I can see, this would create a tight [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) between Zod and your UI framework/library. Doing this, would cause it to be a lot harder to make changes in the future.

If you could provide more context/code to explain why you are wanting to do this, maybe someone would be able to give you some better ideas.

**Reply by winston0410** [reply]
I want to display a message plus a component that when users click on it, apply some transformation to the input value. Right now with string only this is not viable at all. Not all frameworks provide a good support for innerHTML. The framework I am using is Svelte. How do you think I should solve this?

**Reply by JacobWeisenburger** [reply]
Unfortunately, I don't know svelte. Sorry. If it were React or Angular, I would be able to help.

Does anyone else have any ideas?

**Reply by winston0410** [reply]
I think if it is React, you only need to use for jsx for that? something like:

```jsx
{
   message: ( 
  <div>
    "Invalid input"
     <FixButton />
  </div>
  )
}
```

**Reply by scotttrinh** [reply]
I think the closest you're going to get to this is catching errors and parsing them and formatting them appropriately. I realize that's more work and you'd like the schema to be a source of truth for the error display, but that's pretty far out of scope for Zod's core functionality.

You could conceivably turn this "inside out" and define an object that has schemas and error definitions (I think this is roughly what most form-processing libraries do).

**Comment by JacobWeisenburger** [comment, marked as answer]
https://github.com/colinhacks/zod/discussions/877#discussioncomment-2104171
> I think the closest you're going to get to this is catching errors and parsing them and formatting them appropriately. I realize that's more work and you'd like the schema to be a source of truth for the error display, but that's pretty far out of scope for Zod's core functionality.
> 
> You could conceivably turn this "inside out" and define an object that has schemas and error definitions (I think this is roughly what most form-processing libraries do).

~@scotttrinh
